webui/vendor/Bareos/library/Bareos/BSock/BareosBSock.php

```diff
@@ -940,6 +940,73 @@ public function send_command($cmd, $api=0)
         return $result;
     }
 
+    /**
+     * Calls a command on the Bareos Director and returns its result.
+     *
+     * @param $cmd
+     * @return array: Result received from the Bareos Director.
+     */
+    public function call_fullresult_json($cmd)
+    {
+        // should behave similar to
+        // https://github.com/bareos/bareos/blob/master/python-bareos/bareos/bsock/directorconsolejson.py
+        // def call_fullresult(self, command):
+        //     resultstring = super(DirectorConsoleJson, self).call(command)
+        //     data = None
+        //     if resultstring:
+        //         try:
+        //             data = json.loads(resultstring.decode("utf-8"))
+        //         except ValueError as e:
+        //             # in case result is not valid json,
+        //             # create a JSON-RPC wrapper
+        //             data = {"error": {"code": 2, "message": str(e), "data": resultstring}}
+        //             raise bareos.exceptions.JsonRpcInvalidJsonReceivedException(data)
+        //     return data
+        $resultstring = self::send_command($cmd, 2);
+        $data = \Zend\Json\Json::decode($resultstring, \Zend\Json\Json::TYPE_ARRAY);
+        return $data;
+    }
+
+    /**
+     * Calls a command on the Bareos Director and returns its result.
+     *
+     *  If the JSON-RPC result indicates an error
+     *  (contains the ``error`` element),
+     *  an exception will be raised.
+     *
+     * @param $cmd
+     * @return array: The "result" sub-map of the result received from the Bareos Director.
+     * @throw Exception: if JSON-RPC result contains "error" key.
+     * @throw Exception: if JSON-RPC did not contain "result" nor "error" key.
+     */
+    public function call_json($cmd)
+    {
+        // should behave similar to
+        // https://github.com/bareos/bareos/blob/master/python-bareos/bareos/bsock/directorconsolejson.py
+        // def call(self, command):
+        //     json = self.call_fullresult(command)
+        //     if json == None:
+        //         return
+        //     if "result" in json:
+        //         result = json["result"]
+        //     elif "error" in json:
+        //         raise bareos.exceptions.JsonRpcErrorReceivedException(json)
+        //     else:
+        //         raise bareos.exceptions.JsonRpcInvalidJsonReceivedException(json)
+        //     return result
+        $resultstring = self::send_command($cmd, 2);
+        $json = \Zend\Json\Json::decode($resultstring, \Zend\Json\Json::TYPE_ARRAY);
```

Comment:
Why are we not reusing the function we just defined ? Python does the same!

```suggestion
        $json = self::call_fullresult_json($cmd);
```

Reply:
Hmm... should we adapt `call_fullresult_json` so that it detects those conversion errors then ?  As currently we just silently ignore json decode failures (see comment below).
From what i have gathered we can use `json_last_error()` to do so but im not sure what version of php we are targeting.